### DISCUSSION TITLE
NoReturn and object

### DISCUSSION THREAD

**Discussion by JelleZijlstra** [microsoft/pyright#2920] [General]
I was exploring pyright's treatment of NoReturn (in preparation for adding and documenting typing.Never) and noticed that pyright allows passing NoReturn to parameters of type `object`. For example, this typechecks:

```
def f(x: NoReturn):
    repr(x)  # repr() takes object
```

Is this an intentional behavior? It's not what I expected on first principles but it seems useful in practice.

**Comment by erictraut** [comment]
No, that's a bug. Thanks for pointing it out. I'll fix it.

**Reply by JelleZijlstra** [reply]
OK, I'd like to expand on what I said about this behavior being useful in practice. I was looking into using NoReturn more in pyanalyze and got this error:

```
Incompatible argument type for obj: expected object but got NoReturn (code: incompatible_argument)
  Cannot assign NoReturn to object

In /Users/jelle/py/pyanalyze/pyanalyze/signature.py at line 1820
1817:         elif label is ELLIPSIS:
1818:             is_ellipsis = True
1819:         else:
1820:             assert False, repr(label)
                                     ^
1821: 
1822:     return ActualArguments(
1823:         more_processed_args,
```

Here the type checker correctly finds out that the block is dead code (the ifs cover all cases), but it's useful to be able to call `repr(label)` so that the runtime code can give a useful error if the static types are wrong and the else branch does get taken at runtime.

I always write `assert False` for unreachable code this way so I get a useful error message if something unexpected happens at runtime.

**Reply by erictraut** [reply]
Actually, that should be allowed. There is a bug, but it's in the other direction. `NoReturn` should be assignable to any type, not just `object`. That's consistent with how mypy works. It's also consistent with how TypeScript handles its `never` type.

**Reply by JelleZijlstra** [reply]
Oh yes, that's right. The bottom type is assignable to any type but no type is assignable to it.
